File: src/main/java/edu/neu/coe/info6205/union_find/UF.java
package edu.neu.coe.info6205.union_find;

/**
 * This interface models the concept of Union-Find, a special case of the Connections interface
 *
 */
public interface UF extends Connections {
    /**
     * Returns the number of components.
     *
     * @return the number of components (between {@code 1} and {@code n})
     */
    int components();

    /**
     * Returns the component identifier for the component containing site {@code p}.
     *
     * @param  p the integer representing one site
     * @return the component identifier for the component containing site {@code p}
     * @throws IllegalArgumentException unless {@code 0 <= p < n}
     */
    int find(int p);

    /**
     * Merges the component containing site {@code p} with the
     * the component containing site {@code q}.
     *
     * @param  p the integer representing one site
     * @param  q the integer representing the other site
     * @throws IllegalArgumentException unless
     *         both {@code 0 <= p < n} and {@code 0 <= q < n}
     */
    void union(int p, int q);

    /**
     * Returns true if the the two sites are in the same component.
     *
     * @param  p the integer representing one site
     * @param  q the integer representing the other site
     * @return {@code true} if the two sites {@code p} and {@code q} are in the same component;
     *         {@code false} otherwise
     * @throws IllegalArgumentException unless
     *         both {@code 0 <= p < n} and {@code 0 <= q < n}
     */
    default boolean isConnected(int p, int q) {
        return find(p) == find(q);
    }
}

File: Python/src/union_find/uf.py
from abc import abstractmethod

from union_find.connections import Connections


class UF(Connections):

    @abstractmethod
    def components(self) -> int:
        """
        Returns the number of components.

        :return:    the number of components (between 1 and n)
        """
        pass

    @abstractmethod
    def find(self, p: int) -> int:
        """
        Returns the component identifier for the component containing site p.

        :param p:   the integer representing one site
        :return:    the component identifier for the component containing site p
        """
        pass

    @abstractmethod
    def union(self, p: int, q: int) -> None:
        """
        Merges the component containing site p with the
        the component containing site q.

        :param p:   the integer representing one site
        :param q:   the integer representing the other site
        """
        pass

    def is_connected(self, p: int, q: int) -> bool:
        """
        Returns True if the the two sites are in the same component.

        :param p:   the integer representing one site
        :param q:   the integer representing the other site
        :return:    True if the two sites p and q are in the same component;
                    False otherwise
        """
        return self.find(p) == self.find(q)

    @abstractmethod
    def size(self) -> int:
        """
        Returns the number of sites (objects) in this UF object.

        :return:    the number of sites.
        """
        pass
